### PR TITLE
RFC-0019: Proposal for extending pytorch FX Graph Mode Quantization to custom backends

### PR DESCRIPTION
**Pull request by jerryzh168**
will remove https://github.com/pytorch/pytorch/wiki/Extending-PyTorch-Quantization-to-Custom-Backends after this is merged

### CODE REVIEW

File: RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+
+# Proposed APIs for Custom Backend Developers
+```python
+from enum import Enum
+# Demonstration purpose only, the api is subject to change
+class QuantizedOperatorType(Enum):
+    NEED_OBSERVER_FOR_BOTH_INPUTS_AND_OUTPUTS = 0
+    OUTPUT_IS_SHARING_OBSERVER_WITH_INPUT = 1
+
+conv_module_config = {
```
Comment (dzhulgakov) [inline RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md:98]: not sure whether it's the final api, but dataclasses everywhere to make things typed

File: RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+    “type”: QuantizedOperatorType.NEED_OBSERVER_FOR_BOTH_INPUTS_AND_OUTPUTS,
+    “float_to_quantized_operator_mapping”: {
+        # contains mapping for float op and pattern to reference function or reference module
+        “static”: {
+            torch.nn.Conv2d: torch.ao.nn.quantized.reference.Conv2d,
+            (torch.nn.ReLU, torch.nn.Conv2d): torch.ao.nn.quantized.reference.ConvReLU2d,
```
Comment (dzhulgakov) [inline RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md:104]: btw, why do we specify patterns in reverse order, i.e. Relu+Conv2d and not Conv2d+Relu?

File: RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+    # module/functional/torch ops to their configurations
+    “operator”: {
+        torch.nn.Conv2d: conv_module_config,
+        torch.nn.functional.conv2d: conv_functional_config,
+        torch.bmm: bmm_config
+    }
+}
+
+# define a function to return the backend config dict
+def get_custom_backend_config_dict():
```
Comment (dzhulgakov) [inline RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md:138]: and if you make it dataclass, drop `_dict`

File: RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+calibration(model, ...)
+model = convert_fx(model)
+
+# This step will transform a model with reference patterns to a model with
+# fbgemm/qnnpack ops, e.g. torch.ops.quantized.conv2d
+model = lower_to_fbgemm_fx(model)  # lower_to_qnnpack_fx(model)
+```
+
+
+### Example Implementation lower_to_fbgemm_fx
```
Comment (dzhulgakov) [inline RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md:188]: the implementation of lower_to_fbgemm_fx is missing

File: RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+from torch.quantization.quantize_fx import get_backend_config_dict, QuantizedOperatorType, update_operator_config
+import torch.fx
+
+
+# Let's say we want to modify the backend_config of fbgemm and add new quantized operator
+custom_backend_config_dict = get_backend_config_dict("fbgemm")
+bmm_config = {
+    "type": QuantizedOperatorType.NEED_OBSERVER_FOR_BOTH_INPUTS_AND_OUTPUTS,
+}
+update_operator_config(custom_backend_config_dict, torch.bmm, bmm_config)
```
Comment (dzhulgakov) [inline RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md:284]: why `update_operator_config` isn't it just a dict assignment?

**Review by dzhulgakov — APPROVED** [review]

File: RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md (hunk 1 above)
Comment (jerryzh168) [reply RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md:98]: sure, will do that in the end, this is for demo only

File: RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md (hunk 2 above)
Comment (jerryzh168) [reply RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md:104]: we use this pattern to support more complicated subgraphs beyond just a sequence of ops that has single input and single output, here is the doc for the pattern: https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/fx/backend_config_dict/README.md

File: RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md (hunk 5 above)
Comment (jerryzh168) [reply RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md:284]: this is a helper function to hide the complicated nesting in the configuration

File: RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md (hunk 4 above)
Comment (jerryzh168) [reply RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md:188]: oh sorry, should add a `lower_to_fbgemm_fx` that calls `_lower_to_native_backend` below

File: RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md (hunk 5 above)
Comment (jerryzh168) [reply RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md:284]: these are mostly for demo purposes, not final apis

**Pull request merged**